Added non-implace math dunders

FuexFollets opened the pull request.

FuexFollets commented:
It seems fine to include all of these in an example because they are short, but I wonder if I should split some of them. I also wonder if its alright that theres a `__str__` method usage

trag1c reviewed (commented):
I think it would be better to have something more creative, this is just a wrapper around the `int`/`float` types so the only difference is that you use `Number(x)` instead of `x`

trag1c commented:
Also do you think you could implement `__abs__`, `__complex__`, `__pow__`, `__trunc__` too? They're related and it would be weird to have a separate example for them if they're left out.

FuexFollets commented:
As you recommended, I have changed the class representation from `Number` to `Land`. I also added those dunders to the class. I did not want it to be too complicated. Let me know your thoughts.

trag1c commented on src/math-land/non_inplace_math.py:
```diff
@@ -1,5 +1,6 @@
 # Dunders __add__, __sub__, __pos__, __neg__,
 # __mul__, __floordiv__, __truediv__, __divmod__, __mod__, __int__
+# __complex__, __abs__, __trunc__, __pow__
```
Sounds like you missed `__float__` (and maybe some more? didn't check)

trag1c commented:
I think you could also try showing that the `pow()` function can be used for `__pow__` and not just the operator (which is not shown either tbh).

trag1c commented:
Now that I think about it, `__int__`, `__float__` and `__complex__` should probably be part of a different group related to conversions, so I guess you can remove those

FuexFollets commented:
> Now that I think about it, `__int__`, `__float__` and `__complex__` should probably be part of a different group related to conversions, so I guess you can remove those

Ive removed those, although, I think that `__complex__` was the main outlier. The code should still be good at the moment

trag1c commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,112 @@
+
+    def __ceil__(self) -> Land:
+        return Land(math.ceil(self.area))
+
+    def __complex__(self) -> complex:
+        return complex(self.area, 0)
```
Missed this one
```suggestion
```

trag1c commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,114 @@
+    def __mul__(self, scalar: int | float) -> Land:
+        return Land(self.area * scalar)
+
+    def __floordiv__(self, dividend: int | float) -> Land:
+        return Land(self.area // dividend)
+
+    def __truediv__(self, dividend: int | float) -> Land:
+        return Land(self.area / dividend)
+
+    def __divmod__(self, divided_and_modded: int | float) -> tuple[Land, Land]:
+        return (
+            Land(self.area // divided_and_modded),
+            Land(self.area % divided_and_modded),
+        )
+
+    def __mod__(self, modded: int | float) -> Land:
+        return Land(self.area % modded)
```
was the `__divmod__` param supposed to be `divided` or `dividend` lol

FuexFollets replied:
I fixed this variable name with mithic's suggestion.

trag1c commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,114 @@
+
+    print(f"Positive of plot1 area +(plot1.area): {+plot1}")  # Calls __pos__
+    print(f"Negative of plot1 area -(plot1): {-plot1}")  # Calls __neg__
+    print("Absolute value plot2 area: {abs(plot2)}")  # Calls __abs__
```
```suggestion
    print(f"Absolute value plot2 area: {abs(plot2)}")  # Calls __abs__
```

trag1c commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,114 @@
+
+
+class Land:
+    area: int | float = 0
```
Why the class var?

MithicSpirit commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,114 @@
+    def __mul__(self, scalar: int | float) -> Land:
+        return Land(self.area * scalar)
+
+    def __floordiv__(self, dividend: int | float) -> Land:
+        return Land(self.area // dividend)
+
+    def __truediv__(self, dividend: int | float) -> Land:
+        return Land(self.area / dividend)
+
+    def __divmod__(self, divided_and_modded: int | float) -> tuple[Land, Land]:
+        return (
+            Land(self.area // divided_and_modded),
+            Land(self.area % divided_and_modded),
+        )
+
+    def __mod__(self, modded: int | float) -> Land:
+        return Land(self.area % modded)
```
The second argument on all of these (including `__mod__`) should be called `divisor`.

MithicSpirit commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,114 @@
+
+    # Calls __floordiv__
+    print(
+        f"Divisions into plots of land from plot1 of exactly 9 area: "
```
This doesn't need to be an fstring

MithicSpirit reviewed (changes requested):
I think that the example should be changed to be more substantial. To quote James,
> I think the arithmetic dunders should be a number type like in my complex fractions video, inplace versions would naturally be an array or list type

MithicSpirit commented:
I interpreted James's message to mean something more substantial than literally just a wrapper around numbers. The example he gave (complex fraction) has at least a little bit of logic behind it.

Endercheif commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,136 @@
+    )  # Calls __add__
+
+    # Calls __sub__
+    print(
+        f"The value of number_100 taken away from plot 2: {number_100 - number_5point5}"
+    )
+
+    print(
+        f"Positive of number_100 value +(number_100.value): {+number_100}"
+    )  # Calls __pos__
+    print(
+        f"Negative of number_100 value -(number_100): {-number_100}"
+    )  # Calls __neg__
+    print(
+        f"Absolute value number_neg200 value: {abs(number_neg200)}"
+    )  # Calls __abs__
+    print(
+        f"number_100 value being scaled by 3: {number_100 * 3}"
+    )  # Calls __mul__
+
+    # Calls __pow__
+    print(
+        f"number_5point5 value being raised to the power of 3 {number_5point5 ** 3}"
+    )
+    # pow(number_5point5, 3) also works
+
+    # Calls __floordiv__
+    print(
+        "Divisions into plots of land from number_100 of exactly 9 value: "
+        f"{number_100 // 9}"
+    )
+    print(
+        f"Divides number_neg200 into 5 pieces: {number_neg200 / 5}"
+    )  # Calls __truediv__
+
+    divisions, remainder = divmod(number_100, 7)  # Calls __divmod__
+    print(
+        f"number_100 can be divided into {divisions} plots of 7 value each "
+        f"with a remainder of {remainder} value"
+    )
```
Consistency with comments 

```suggestion
    # Calls __add__
    print(
        f"Combined value of number_100, number_neg200: {number_100 + number_5point5}"
    ) 

    # Calls __sub__
    print(
        f"The value of number_100 taken away from plot 2: {number_100 - number_5point5}"
    )

    # Calls __pos__
    print(
        f"Positive of number_100 value +(number_100.value): {+number_100}"
    )

    # Calls __neg__
    print(
        f"Negative of number_100 value -(number_100): {-number_100}"
    )

    # Calls __abs__
    print(
        f"Absolute value number_neg200 value: {abs(number_neg200)}"
    )

    # Calls __mul__
    print(
        f"number_100 value being scaled by 3: {number_100 * 3}"
    )

    # Calls __pow__
    print(
        f"number_5point5 value being raised to the power of 3 {number_5point5 ** 3}"
    )
    # pow(number_5point5, 3) also works

    # Calls __floordiv__
    print(
        "Divisions into plots of land from number_100 of exactly 9 value: "
        f"{number_100 // 9}"
    )

    # Calls __truediv__
    print(
        f"Divides number_neg200 into 5 pieces: {number_neg200 / 5}"
    )

    # Calls __divmod__
    divisions, remainder = divmod(number_100, 7)  
    print(
        f"number_100 can be divided into {divisions} plots of 7 value each "
        f"with a remainder of {remainder} value"
    )
```

Endercheif reviewed (changes requested):
Looks good. I feel like it would have been better if it was a bit more of a creative example as it's mostly just a wrapper for `int`/`float`

FuexFollets commented:
> Looks good. I feel like it would have been better if it was a bit more of a creative example as it's mostly just a wrapper for `int`/`float`

I did have that before, maybe a fraction would have been a better example, but this works just fine

trag1c commented:
> I did have that before, maybe a fraction would have been a better example, but this should work just fine

This is a literal number wrapper which, as James said, isn't too substantial. A complex fraction or even a regular fraction (to maybe show how built-in `fractions.Fraction` works? 🙂) example would be nice.

FuexFollets commented:
> > I did have that before, maybe a fraction would have been a better example, but this should work just fine
> 
> This is a literal number wrapper which, as James said, isn't too substantial. A complex fraction or even a regular fraction (to maybe show how built-in `fractions.Fraction` works? slightly_smiling_face) example would be nice.

Sure, lemme rewrite everything

trag1c commented:
Thank you :pray:

FuexFollets commented:
Btw, I will end up needing to also implement __ceil__, __floor__, __int__, __float__ for this one

CircuitSacul commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,189 @@
+
+class Fraction:
+    def __init__(self, numerator: int, denominator: int) -> None:
+        self.numerator: int = numerator
+        self.denominator: int = denominator
```
These typehints are redundant

CircuitSacul commented on src/math-land/non_inplace_math.py:
```diff
@@ -0,0 +1,189 @@
+        self.denominator: int = denominator
+
+    def simplify(self) -> Fraction:
+        gcd: int = math.gcd(self.numerator, self.denominator)
```
Another redundant typehint

CircuitSacul reviewed (approved):
Redundant typehints are kinda a minor thing, so consider this an approval anyways.

This file reminds me of that time I wrote my own fractions lib only to discover `fractions` was in the stdlib.

trag1c commented:
Make black happy :+1:

qexat reviewed (approved):
Besides what @CircuitSacul said, LGTM
